test/csuite/wt12015_backup_corruption/main.c

```diff
@@ -339,7 +339,8 @@ run_test_backup(void)
     /* Parent. */
 
     /* Wait for the child to die. */
-    testutil_assert(waitpid(pid, &status, 0) > 0);
+    ret = waitpid(pid, &status, 0);
+    testutil_assert(ret != 0 && WIFSIGNALED(status) != 0);
```

Comment:
Looking at the documentation, this means: "The call is interrupted by a caught signal or the signal does not have the SA_RESTART flag set."

I wonder if this is because we set a child death handler earlier, and it is thus expected? If that's the case, we should write the assert accordingly, and add a comment explaining it.

Reply:
On my Mac, `status` was 0 at the end of the call, so I don't think we can rely on it. If `waitpid` fails, the `status` variable is not guaranteed to be set—it could be the previous value, and as the function does not actually initialize the value of `status`, it could just contain some random data.

Reply:
Doing more experimentation and after more documentation reading, I believe this is the right answer:
```
testutil_assert(waitpid(pid, &status, 0) > 0 || errno == EINTR || errno == ECHILD);
```

Ideally, `waitpid` would return a positive number, which is the PID. This is what normally happens on Linux. According to the documentation `EINTR` is an expected error in the case `SIGCHLD` is caught, and we are indeed setting up our own `SIGCHLD` handler, so this makes sense.

I'm not entirely convinced that we need to check for `ECHILD`. This could happen if the PID is not the child process (which could theoretically happen, but the probability that the child exits and the PID gets recycled in the meantime is minuscule) or with certain `SIGCHLD` signal handler settings (which I don't think we use), but I think it would be okay to nonetheless include.

Reply:
Do we need to check things here at all? If it is this variable and this is not the important part of the test. The only thing that seemed consistent between systems was `WIFSIGNALED(status)` was true. Did you not see that on your Mac? Otherwise I'd be inclined to ignore the child status and just test the database contents.

Reply:
That's right, status was 0 on my setup. If the function returns an error, setting the value of status is not guaranteed, so we can't rely on that behavior. 

From reading the documentation, waitpid on Linux is also allowed to fail with EINTR in this case, so that much more we should explicitly allow it.

I think it's a good idea to check for return codes so that we know if waitpid is misbehaving in some strange way in the future; I don't expect it, but it would be hard to debug otherwise.

Reply:
`waitpid()` definitely returns the *if* there is a child PID to wait upon, it will return zero if there are no eligible children and there was no error.

Linux waitpid(2):
> ... if WNOHANG was specified and one or more child(ren) specified by pid exist, but have not yet changed state, then 0 is returned.

MacOS WAIT(2):
> ... Otherwise, if WNOHANG is specified and there are no stopped or exited children, 0 is returned.